src/components/views/rooms/Autocomplete.js

```diff
@@ -42,38 +54,67 @@ export default class Autocomplete extends React.Component {
         });
     }
 
-    onUpArrow() {
-        this.setState({selectionOffset: this.state.selectionOffset - 1});
+    countCompletions(): number {
+        return this.state.completions.map(completionResult => {
+            return completionResult.completions.length;
+        }).reduce((l, r) => l + r);
+    }
+
+    // called from MessageComposerInput
+    onUpArrow(): boolean {
+        let completionCount = this.countCompletions(),
+            selectionOffset = (completionCount + this.state.selectionOffset - 1) % completionCount;
+        this.setSelection(selectionOffset);
         return true;
     }
 
-    onDownArrow() {
-        this.setState({selectionOffset: this.state.selectionOffset + 1});
+    // called from MessageComposerInput
+    onDownArrow(): boolean {
+        let completionCount = this.countCompletions(),
+            selectionOffset = (this.state.selectionOffset + 1) % completionCount;
+        this.setSelection(selectionOffset);
         return true;
     }
 
+    /** called from MessageComposerInput
+     * @returns {boolean} whether confirmation was handled
+     */
+    onConfirm(): boolean {
+        if (this.countCompletions() === 0)
```

Comment:
We should write up a code style doc, but we settled on either single-line if statements or braces if the clause goes on its own line